[Homework Questions Link](https://github.com/DataTalksClub/stock-markets-analytics-zoomcamp/blob/main/cohorts/2026/homework2.md)

## Question 1: [IPO] Withdrawn IPOs by Company Type

### What is the total withdrawn IPO value (in $ millions) for the company class with the highest total withdrawal value?

In [77]:
import pandas as pd
import re
import numpy as np

In [11]:
url = "https://www.iposcoop.com/ipos-recently-filed/"
ipos_recently_filed = pd.read_html(url)[0]

# look at the ipo data
ipos_recently_filed.head(3)

,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating
0,2026-09-28,Encore Medical Inc.,EMI,The Oak Ridge Financial Services Group/Network...,3.0,$5.00,$5.00,$15.00,Wednesday,S/O
1,2026-09-25,Elevation Acquisition Group,ELEVU,Maxim Group,10.0,$10.00,$10.00,$100.00,TBA,S/O
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.0,$15.00,$18.00,$825.00,Withdrawn,S/O


In [12]:
ipos_recently_filed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   File Date             500 non-null    object 
 1   Company               500 non-null    object 
 2   Symbol                500 non-null    object 
 3   Managers              500 non-null    object 
 4   Shares (millions)     500 non-null    float64
 5   Price Low             464 non-null    object 
 6   Price High            464 non-null    object 
 7   Est $ Vol (millions)  495 non-null    object 
 8   Expected To Trade     500 non-null    object 
 9   SCOOP Rating          500 non-null    object 
dtypes: float64(1), object(9)
memory usage: 39.2+ KB


In [71]:
# filter to 'Withdrawn' only for Expected to Trade
withdrawn_df = ipos_recently_filed[ipos_recently_filed['Expected To Trade']=='Withdrawn']
display(withdrawn_df.shape)

(35, 10)

In [72]:
# classifies company type based on keywords
def classify_company_type(company_name):
    company_name_lower = str(company_name).lower()
    if 'technologies' in company_name_lower:
        return 'Technologies'
    elif 'acquisition corp' in company_name_lower or \
         'acquisition corporation' in company_name_lower or \
         re.search(r'\bcorp\b', company_name_lower):
        return 'Acquisition Corp'
    elif 'inc' in company_name_lower or 'incorporated' in company_name_lower:
        return 'Inc.'
    elif 'group' in company_name_lower:
        return 'Group'
    elif 'ltd' in company_name_lower or 'limited' in company_name_lower:
        return 'Limited'
    elif 'holdings' in company_name_lower or 'holding' in company_name_lower:
        return 'Holdings'
    else:
        return 'Other'

withdrawn_df['Company Type'] = withdrawn_df['Company'].apply(classify_company_type)
withdrawn_df.head(3)

/tmp/ipykernel_1996/1360166711.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Company Type'] = withdrawn_df['Company'].apply(classify_company_type)


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,$825.00,Withdrawn,S/O,Other
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,$7.20,Withdrawn,S/O,Group
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,$16.88,Withdrawn,S/O,Group


In [73]:
def find_avg_price(row: pd.DataFrame):
  low = pd.to_numeric(str(row["Price Low"]).replace('$', ''), errors="coerce")
  high = pd.to_numeric(str(row["Price High"]).replace('$', ''), errors="coerce")

  return (low + high) / 2

withdrawn_df['Avg_price'] = withdrawn_df.apply(find_avg_price, axis=1)
withdrawn_df.head(3)

/tmp/ipykernel_1996/597125063.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Avg_price'] = withdrawn_df.apply(find_avg_price, axis=1)


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,$825.00,Withdrawn,S/O,Other,16.5
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,$7.20,Withdrawn,S/O,Group,4.0
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,$16.88,Withdrawn,S/O,Group,4.5


In [74]:
withdrawn_df["Shares (millions)"] = pd.to_numeric(withdrawn_df['Shares (millions)'], errors="coerce")
withdrawn_df['Est $ Vol (millions)'] = pd.to_numeric(withdrawn_df['Est $ Vol (millions)'].str.replace('$', '').replace(',', ''), errors="coerce")
withdrawn_df.head(3)

/tmp/ipykernel_1996/761777133.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df["Shares (millions)"] = pd.to_numeric(withdrawn_df['Shares (millions)'], errors="coerce")
/tmp/ipykernel_1996/761777133.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Est $ Vol (millions)'] = pd.to_numeric(withdrawn_df['Est $ Vol (millions)'].str.replace('$', '').replace(',', ''), errors="coerce")


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,825.00,Withdrawn,S/O,Other,16.5
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,7.20,Withdrawn,S/O,Group,4.0
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,16.88,Withdrawn,S/O,Group,4.5


In [79]:
withdrawn_df['Shares_offered_value'] = np.where(
    pd.isna(withdrawn_df['Shares (millions)'] * withdrawn_df['Avg_price']),
    withdrawn_df['Est $ Vol (millions)'],
    withdrawn_df['Shares (millions)'] * withdrawn_df['Avg_price']
)
withdrawn_df.head()

/tmp/ipykernel_1996/804680285.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Shares_offered_value'] = np.where(


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price,Shares_offered_value
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,825.00,Withdrawn,S/O,Other,16.5,825.000
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,7.20,Withdrawn,S/O,Group,4.0,7.200
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,16.88,Withdrawn,S/O,Group,4.5,16.875
39,2026-09-15,OTSAW Ltd. (Withdrawn),OTSA,Aegis Capital,4.40,$4.50,$5.50,22.00,Withdrawn,S/O,Limited,5.0,22.000
45,2026-09-10,Motive Technologies (Withdrawn),MTVE,J.P.Morgan/Citigroup/Barclays/Jefferies/RBC Ca...,0.00,NaN,NaN,100.00,Withdrawn,S/O,Technologies,NaN,100.000


In [92]:
agg_df = withdrawn_df.groupby(['Company Type']) \
.Shares_offered_value.agg(['sum']) \
.rename({'sum': 'sum_shares_offered_value'}, axis=1)
agg_df

,sum_shares_offered_value
Company Type,
Acquisition Corp,499.9850
Group,56.5750
Holdings,311.6575
Inc.,351.0000
Limited,219.2500
Other,1115.4450
Technologies,184.9000


In [90]:
agg_df[agg_df.sum_shares_offered_value == agg_df.sum_shares_offered_value.max()]

,sum_shares_offered_value
Company Type,
Other,1115.445


Answer: The class 'Other' has the highest total value of withdrawals with $1,115.445.